### ISSUE TITLE
Handling additional response_types

### ISSUE THREAD

**Issue by coco1979ka** [spring-projects/spring-security-oauth#619]
Currently Spring Security OAuth supports code and token as response type. If I want to implement OpenID Connect Protocol I have to allow hybrid response types (e.g. response_type=code id_token). 

I wonder how I can extend the framework to support this. Should I really try to replace the AuthorizationEndpoint with my own implementation? Is this possible at all? (the EndpointsConfigurer does not allow to set authorizationEndpoint).

**Comment by oharsta** [comment]
See #626

**Comment by coco1979ka** [comment]
@oharsta after three weeks still no response... Seems to me, that we should think about a pull request. But when would it be merged? Sometimes this project seems to be inactive...

**Comment by oharsta** [comment]
I have already written the pull request 9 days ago. But I did not receive any comments.See https://github.com/spring-projects/spring-security-oauth/pull/627. 

I'm using that forked code in my own project and it works for me. See https://github.com/OpenConext/OpenConext-oidc/blob/master/oidc-server/src/main/java/oidc/service/OpenIdResponseTypesHandler.java

**Comment by dsyer** [comment]
We do want to support custom response types, but I'm not sure how and when. The pull request looks like it is complete, tested and supports the existing default behaviour, so it is close to being applicable if we decide we want this feature on master (i.e. 2.0.9 likely). A picky comment:

I don't like the name of the new strategy: "custom" seems like it should be unnecessary - ideally there should only be a strategy with a default implementation that handles "token" and "code" responses. Customizations can then provide an implementation of the strategy.

Sorry about the delay: there isn't enough bandwidth in the Spring team to cover this project day to day, but we are trying to hire someone to help. If you know anyone who wants a job doing this then tell them to contact me directly.

**Comment by oharsta** [comment]
Thanks for your reply. I totally agree with your comment, but the pull request is already relatively large and implementing your suggestion would mean a complete overhaul of the AuthorizationEndpoint (and AuthorizationServerEndpointsConfiguration#authorizationEndpoint

Do you expect a new squashed pull request with your suggestion or is this good to go if you decide you want this feature?

**Comment by dsyer** [comment]
The default response handler would contain lot of code from the existing `AuthorizationEndpoint` you mean? Is that why it ends up being an overhaul? Maybe we can preserve the existing APIs though?

I can't really say until we see the code, but I suspect that my preferred approach (as outlined above) will be too large of a change for 2.0.9, which is what I suspected anyway. If you want to try doing that so we can get a feel for it that would be excellent. I suppose maybe a new PR would be better (up to you though).

**Comment by oharsta** [comment]
I have made a new pull request with a ```DefaultResponseTypesHandler``` implementation of the ```ResponseTypesHandler```. The branch is based on the original pull request and I have merged the latest changes from upstream master.

https://github.com/spring-projects/spring-security-oauth/pull/657

Ideally the ```AuthorizationCodeServices``` implementation should be configured in the ```DefaultResponseTypesHandler```, but then I can't override when the ```registerAuthorizationEndpoint``` is false. Either therefore I have to pass a reference of the ```AuthorizationCodeServices``` into the ```ResponseTypesHandler#handleApprovedAuthorizationRequest``` or add a setter for
```AuthorizationCodeServices``` in the ```ResponseTypesHandler``` interface to set the ```AuthorizationCodeServices``` when the user overrides the default. I have chosen the first option.

Also the tests now do not reflect the new class composition of the ```AuthorizationEndpoint```. As the new composition is 100% backward compatible with the defaults this is ok for test coverage (it has not become less). Changing and moving all tests around makes it much more difficult to track the changes I have made. However this can be a future enhancement when this pull request gets accepted.

Any comments are more than welcome as this turned out to be (as I predicted) a large refactor exercise.
